linkerd/cache/src/lib.rs

```diff
@@ -19,32 +21,117 @@ where
 {
     inner: N,
     services: Services<T, S>,
+
+    // As long as this is held, the eviction tasks will activate every time it is
+    // notified.
+    evict: Arc<Notify>,
 }
 
 /// A tracker inserted into each inner service that, when dropped, indicates the
 /// service may be removed from the cache.
-pub type Handle = Arc<()>;
+#[derive(Debug)]
+pub struct Handle {
+    inner: Arc<()>,
+    evict: Weak<Notify>,
+}
 
 type Services<T, S> = Arc<RwLock<HashMap<T, (S, Weak<()>)>>>;
 
+// === impl Handle ===
+
+impl Drop for Handle {
+    fn drop(&mut self) {
+        if let Some(evict) = self.evict.upgrade() {
+            evict.notify_one();
+        }
+    }
+}
+
 // === impl Cache ===
 
 impl<T, N> Cache<T, N, N::Service>
 where
-    T: Eq + Hash,
-    N: NewService<(Handle, T)>,
+    T: Eq + Hash + Send + Sync + 'static,
+    N: NewService<(Handle, T)> + 'static,
+    N::Service: Send + Sync + 'static,
 {
     pub fn layer() -> impl layer::Layer<N, Service = Self> + Copy + Clone {
-        layer::mk(|inner| Self {
-            inner,
-            services: Services::default(),
+        layer::mk(|inner| {
+            let (cache, _) = Self::spawn(inner);
+            cache
         })
     }
 
-    fn new_entry(new: &mut N, target: T) -> (N::Service, Weak<()>) {
-        let handle = Arc::new(());
-        let weak = Arc::downgrade(&handle);
-        let svc = new.new_service((handle, target));
+    fn spawn(inner: N) -> (Self, tokio::task::JoinHandle<()>) {
+        let services = Services::default();
+        let evict = Arc::new(Notify::new());
+
+        // Spawn a background task to remove defunct services.
+        let task = tokio::spawn(
+            Self::evict(Arc::downgrade(&evict), services.clone()).instrument(debug_span!("evict")),
```

Comment:
nit/tioli: could also just stick an `#[instrument(level = "debug", skip(evict, services)]` attribute on `evict`...about the same number of characters i guess...